File: server-rs/src/db/rules.rs
use anyhow::Result;
use rusqlite::params;
use uuid::Uuid;

use crate::models::RepositoryRule;

use super::{Db, now_iso};

impl Db {
    /// List rules for prompt injection: repo-specific + global rules.
    pub fn list_rules_for_prompt(&self, repo_id: &str) -> Result<Vec<RepositoryRule>> {
        let conn = self.connect()?;
        let mut stmt = conn.prepare(
            "SELECT id, repo_id, content, source, source_comment_id, created_at, updated_at
             FROM repository_rules
             WHERE repo_id = ?1 OR repo_id IS NULL
             ORDER BY created_at ASC",
        )?;
        let rows = stmt.query_map(params![repo_id], |row| {
            Ok(RepositoryRule {
                id: row.get(0)?,
                repo_id: row.get(1)?,
                content: row.get(2)?,
                source: row.get(3)?,
                source_comment_id: row.get(4)?,
                created_at: row.get(5)?,
                updated_at: row.get(6)?,
            })
        })?;
        rows.collect::<std::result::Result<Vec<_>, _>>()
            .map_err(anyhow::Error::from)
    }

    /// List rules for UI. If repo_id is given, returns repo + global. Otherwise global only.
    pub fn list_rules(&self, repo_id: Option<&str>) -> Result<Vec<RepositoryRule>> {
        if let Some(rid) = repo_id {
            self.list_rules_for_prompt(rid)
        } else {
            let conn = self.connect()?;
            let mut stmt = conn.prepare(
                "SELECT id, repo_id, content, source, source_comment_id, created_at, updated_at
                 FROM repository_rules
                 WHERE repo_id IS NULL
                 ORDER BY created_at ASC",
            )?;
            let rows = stmt.query_map([], |row| {
                Ok(RepositoryRule {
                    id: row.get(0)?,
                    repo_id: row.get(1)?,
                    content: row.get(2)?,
                    source: row.get(3)?,
                    source_comment_id: row.get(4)?,
                    created_at: row.get(5)?,
                    updated_at: row.get(6)?,
                })
            })?;
            rows.collect::<std::result::Result<Vec<_>, _>>()
                .map_err(anyhow::Error::from)
        }
    }

    pub fn create_rule(
        &self,
        repo_id: Option<&str>,
        content: &str,
        source: &str,
        source_comment_id: Option<&str>,
    ) -> Result<RepositoryRule> {
        let conn = self.connect()?;
        let id = Uuid::new_v4().to_string();
        let ts = now_iso();
        conn.execute(
            "INSERT INTO repository_rules (id, repo_id, content, source, source_comment_id, created_at, updated_at)
             VALUES (?1, ?2, ?3, ?4, ?5, ?6, ?7)",
            params![id, repo_id, content, source, source_comment_id, ts, ts],
        )?;
        Ok(RepositoryRule {
            id,
            repo_id: repo_id.map(String::from),
            content: content.to_string(),
            source: source.to_string(),
            source_comment_id: source_comment_id.map(String::from),
            created_at: ts.clone(),
            updated_at: ts,
        })
    }

    pub fn update_rule(&self, id: &str, content: &str) -> Result<()> {
        let conn = self.connect()?;
        let ts = now_iso();
        let n = conn.execute(
            "UPDATE repository_rules SET content = ?1, updated_at = ?2 WHERE id = ?3",
            params![content, ts, id],
        )?;
        if n == 0 {
            return Err(anyhow::anyhow!("Rule not found"));
        }
        Ok(())
    }

    pub fn delete_rule(&self, id: &str) -> Result<()> {
        let conn = self.connect()?;
        conn.execute("DELETE FROM repository_rules WHERE id = ?1", [id])?;
        Ok(())
    }

    pub fn get_rule_by_id(&self, id: &str) -> Result<Option<RepositoryRule>> {
        let conn = self.connect()?;
        conn.query_row(
            "SELECT id, repo_id, content, source, source_comment_id, created_at, updated_at
             FROM repository_rules WHERE id = ?1",
            [id],
            |row| {
                Ok(RepositoryRule {
                    id: row.get(0)?,
                    repo_id: row.get(1)?,
                    content: row.get(2)?,
                    source: row.get(3)?,
                    source_comment_id: row.get(4)?,
                    created_at: row.get(5)?,
                    updated_at: row.get(6)?,
                })
            },
        )
        .optional()
        .map_err(anyhow::Error::from)
    }

    /// Bulk replace rules for a given scope (repo_id or global).
    /// Deletes existing rules for that scope and inserts new ones.
    pub fn bulk_replace_rules(
        &self,
        repo_id: Option<&str>,
        contents: &[String],
    ) -> Result<Vec<RepositoryRule>> {
        let mut conn = self.connect()?;
        let tx = conn.transaction()?;
        let ts = now_iso();

        if let Some(rid) = repo_id {
            tx.execute(
                "DELETE FROM repository_rules WHERE repo_id = ?1",
                params![rid],
            )?;
        } else {
            tx.execute(
                "DELETE FROM repository_rules WHERE repo_id IS NULL",
                [],
            )?;
        }

        let mut result = Vec::with_capacity(contents.len());
        for content in contents {
            let id = Uuid::new_v4().to_string();
            tx.execute(
                "INSERT INTO repository_rules (id, repo_id, content, source, source_comment_id, created_at, updated_at)
                 VALUES (?1, ?2, ?3, 'manual', NULL, ?4, ?5)",
                params![id, repo_id, content, ts, ts],
            )?;
            result.push(RepositoryRule {
                id,
                repo_id: repo_id.map(String::from),
                content: content.clone(),
                source: "manual".to_string(),
                source_comment_id: None,
                created_at: ts.clone(),
                updated_at: ts.clone(),
            });
        }

        tx.commit()?;
        Ok(result)
    }
}

use rusqlite::OptionalExtension;

File: server-ts/src/db/rules.ts
import { v4 as uuidv4 } from 'uuid';
import type { RepositoryRule } from '../models.js';
import { Db, nowIso } from './index.js';

declare module './index.js' {
  interface Db {
    listRulesForPrompt(repoId: string): RepositoryRule[];
    listRules(repoId?: string): RepositoryRule[];
    createRule(
      repoId: string | undefined,
      content: string,
      source: string,
      sourceCommentId?: string,
    ): RepositoryRule;
    updateRule(id: string, content: string): void;
    deleteRule(id: string): void;
    getRuleById(id: string): RepositoryRule | null;
    bulkReplaceRules(repoId: string | undefined, contents: string[]): RepositoryRule[];
  }
}

const RULE_COLS =
  'id, repo_id, content, source, source_comment_id, created_at, updated_at';

function rowToRule(row: any): RepositoryRule {
  return {
    id: row.id,
    repo_id: row.repo_id,
    content: row.content,
    source: row.source,
    source_comment_id: row.source_comment_id,
    created_at: row.created_at,
    updated_at: row.updated_at,
  };
}

Db.prototype.listRulesForPrompt = function (repoId: string): RepositoryRule[] {
  const db = this.connect();
    const rows = db
      .prepare(
        `SELECT ${RULE_COLS} FROM repository_rules
         WHERE repo_id = ? OR repo_id IS NULL
         ORDER BY created_at ASC`,
      )
      .all(repoId) as any[];
    return rows.map(rowToRule);
};

Db.prototype.listRules = function (repoId?: string): RepositoryRule[] {
  if (repoId) {
    return this.listRulesForPrompt(repoId);
  }

  const db = this.connect();
    const rows = db
      .prepare(
        `SELECT ${RULE_COLS} FROM repository_rules
         WHERE repo_id IS NULL
         ORDER BY created_at ASC`,
      )
      .all() as any[];
    return rows.map(rowToRule);
};

Db.prototype.createRule = function (
  repoId: string | undefined,
  content: string,
  source: string,
  sourceCommentId?: string,
): RepositoryRule {
  const db = this.connect();
    const id = uuidv4();
    const ts = nowIso();
    db.prepare(
      `INSERT INTO repository_rules (id, repo_id, content, source, source_comment_id, created_at, updated_at)
       VALUES (?, ?, ?, ?, ?, ?, ?)`,
    ).run(id, repoId ?? null, content, source, sourceCommentId ?? null, ts, ts);

    return {
      id,
      repo_id: repoId ?? null,
      content,
      source,
      source_comment_id: sourceCommentId ?? null,
      created_at: ts,
      updated_at: ts,
    };
};

Db.prototype.updateRule = function (id: string, content: string): void {
  const db = this.connect();
    const result = db
      .prepare('UPDATE repository_rules SET content = ?, updated_at = ? WHERE id = ?')
      .run(content, nowIso(), id);
    if (Number(result.changes) === 0) {
      throw new Error('Rule not found');
    }
};

Db.prototype.deleteRule = function (id: string): void {
  const db = this.connect();
    db.prepare('DELETE FROM repository_rules WHERE id = ?').run(id);
};

Db.prototype.getRuleById = function (id: string): RepositoryRule | null {
  const db = this.connect();
    const row = db
      .prepare(`SELECT ${RULE_COLS} FROM repository_rules WHERE id = ?`)
      .get(id) as any | undefined;
    return row ? rowToRule(row) : null;
};

Db.prototype.bulkReplaceRules = function (
  repoId: string | undefined,
  contents: string[],
): RepositoryRule[] {
  const db = this.connect();
    const ts = nowIso();
    const result: RepositoryRule[] = [];

    const tx = this.transaction(() => {
      if (repoId) {
        db.prepare('DELETE FROM repository_rules WHERE repo_id = ?').run(repoId);
      } else {
        db.prepare('DELETE FROM repository_rules WHERE repo_id IS NULL').run();
      }

      for (const content of contents) {
        const id = uuidv4();
        db.prepare(
          "INSERT INTO repository_rules (id, repo_id, content, source, source_comment_id, created_at, updated_at) VALUES (?, ?, ?, 'manual', NULL, ?, ?)",
        ).run(id, repoId ?? null, content, ts, ts);

        result.push({
          id,
          repo_id: repoId ?? null,
          content,
          source: 'manual',
          source_comment_id: null,
          created_at: ts,
          updated_at: ts,
        });
      }
    });
    tx();

    return result;
};
